### PR TITLE
git_backend: replace git2::Repository with gix::Repository

### PR DESCRIPTION
**Pull request by yuja**
#2316 

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by PhilipMetzger** [comment]
Thanks for doing this, it's nice to see that we can intergrate Gix so easily. 

Can we protect the different Git backends with a runtime flag, as I think a feature is overkill? E.g `git.backend = "gix"` or `git.backend = "git2"`. 

> tests: force gitoxide to not load config nor use "main" as default branch
> 
> AFAIK, there's no global config state for gitoxide. We can use
> Config::isolated() in tests, but GitBackend should load config files in a
> normal way.

Can we file a upstream bug for more configuration options and see what happens?  

I think martin should do the final review though, as the Git backend is not my expertise.

**Comment by PhilipMetzger** [comment]
> That's doable, but I don't think we'll need runtime or compile-time flag. They're functionally equivalent, and having separate backends will double the maintenance cost.

I agree that this doubles the maintenance but before we finish the migration to Gix it would be wise to allow a partial rollout.  

> gitoxide provides various ways to configure repository instance. It just doesn't have a global switch. I think that's by design. Perhaps, a better workaround is to initialize `GitBackend` with a specific configuration.

Maybe we'll need a config shim until libgit2 is gone.

**Comment by yuja** [comment]
> > That's doable, but I don't think we'll need runtime or compile-time flag. They're functionally equivalent, and having separate backends will double the maintenance cost.
> 
> I agree that this doubles the maintenance cost but before we finish the migration to Gix it would be wise to allow a partial rollout.

We won't fully drop the libgit2 dependency anytime soon (because push/fetch API isn't mature afaik.) If we think it's too early to migrate "some" of the git interface to gitoxide, it's probably better to stick to libgit2.

**Comment by martinvonz** [comment]
My main concern is that we might run into some bug in gitoxide that makes it unable to access certain repositories (maybe using some rarely used packfile feature or running into some corner case in the packfile format). I'm going to cut jj 0.11 today, so right after that seems like the perfect time to switch. Then we'll have a full month to test it. Does that sound safe enough, @PhilipMetzger?

**Comment by PhilipMetzger** [comment]
> We won't fully drop the libgit2 dependency anytime soon (because push/fetch API isn't mature afaik.) If we think it's too early to migrate "some" of the git interface to gitoxide, it's probably better to stick to libgit2.

Yes, I'm aware that this won't be anytime soon, but when they have feature parity a flag would make sense. 

> My main concern is that we might run into some bug in gitoxide that makes it unable to access certain repositories (maybe using some rarely used packfile feature or running into some corner case in the packfile format).

Yes, that perfectly encapsulates my concerns as well. I just don't want [Hyrum] to show up. Every "identical" implementation must be bug for bug compatible to not fall victim to it.  

> Then we'll have a full month to test it. Does that sound safe enough, @PhilipMetzger?

Seems great, we'll what happens. 

[Hyrum]: https://www.hyrumslaw.com/

**Comment by martinvonz** [comment]
@Byron, I thought you might be interested in this PR

**Comment by Byron** [comment]
Thanks for reeling me in! I will definitely take a look once you think it's nearing completion to see if I *like* the `gix` usage especially compared to `git2`, and if not can make improvements.

> My main concern is that we might run into some bug in gitoxide that makes it unable to access certain repositories (maybe using some rarely used packfile feature or running into some corner case in the packfile format).

Even though `gix` is actually more compatible than `git2` when it comes to handling local repositories and when fetching, there is currently [one known and expensive-to-overcome limitation](https://github.com/Byron/gitoxide/issues/1025) that makes fetching from azure-repos fail.
Push isn't supported at all, and when handling `file://` and `ssh://` URLs it will shell out to `git` and `ssh` respectively, thus isn't as self-contained as `git2` is. All this will be fixed eventually as self-containment is also required for `cargo`.

What would help me is to have an issue (maybe over at `gitoxide`) that has a check-list of feature that are present and missing. Then I could even let you know once the next level of integration can be reached. My feeling is that when `cargo` is integrated, and `push` is available on top of that, you would be good here to finalize the switch. And `push` *should* be nearly free once support for `file://` is built-in.

**Comment by martinvonz** [comment]
> Thanks for reeling me in! I will definitely take a look once you think it's nearing completion to see if I _like_ the `gix` usage especially compared to `git2`, and if not can make improvements.

It's actually already complete :) However, this is just for jj's "git commit backend", which just interacts with the object store (read and write commits, trees, blobs).

> > My main concern is that we might run into some bug in gitoxide that makes it unable to access certain repositories (maybe using some rarely used packfile feature or running into some corner case in the packfile format).
> 
> Even though `gix` is actually more compatible than `git2` when it comes to handling local repositories and when fetching, there is currently [one known and expensive-to-overcome limitation](https://github.com/Byron/gitoxide/issues/1025) that makes fetching from azure-repos fail.

Interesting! I think ref-deltas was what I was thinking of as a feature gitoxide might not support well (I had heard about them from the Git team at Google), though I didn't even remember how they worked.

> Push isn't supported at all, and when handling `file://` and `ssh://` URLs it will shell out to `git` and `ssh` respectively, thus isn't as self-contained as `git2` is. All this will be fixed eventually as self-containment is also required for `cargo`.
> 
> What would help me is to have an issue (maybe over at `gitoxide`) that has a check-list of feature that are present and missing. Then I could even let you know once the next level of integration can be reached. My feeling is that when `cargo` is integrated, and `push` is available on top of that, you would be good here to finalize the switch. And `push` _should_ be nearly free once support for `file://` is built-in.

Sounds good. We'll continue to use libgit2 for push and fetch for now.

**Comment by martinvonz** [comment]
@Byron, just to be sure, the `REF_DELTA` limitation is just for intra-pack references, right? That is, is gitoxide able to handle `REF_DELTA`s pointing to an already existing pack?

**Review by martinvonz — APPROVED** [review]

File: lib/testutils/src/lib.rs
Comment (martinvonz) [inline lib/testutils/src/lib.rs:1]: > GitBackend should load config files in a normal way

Do you know what difference it makes in practice? Is it things like enabling new format-related features (like the commit graph, perhaps)?

File: lib/src/git_backend.rs [hunk 1]
```diff
@@ -93,7 +91,12 @@ impl From<GitBackendError> for BackendError {
 }
 
 pub struct GitBackend {
-    repo: Mutex<git2::Repository>,
+    // While gix::Repository can be created from gix::ThreadSafeRepository, it's
+    // cheaper to cache the thread-local instance behind a mutex than creating
+    // one for each backend method call. Our GitBackend is most likely to be
+    // used in a single-threaded context.
```
Comment (martinvonz) [inline lib/src/git_backend.rs:97]: Maybe `gix::ThreadSafeRepository` would be used in `ReadonlyGitBackend` and `gix::Repository` would be used in `MutableGitBackend` if we ever implement #50?

File: lib/testutils/src/lib.rs
Comment (yuja) [reply lib/testutils/src/lib.rs:1]: I haven't checked which option is read by gitoxide and affects our use case, but yes, `core.commitGraph` appears to be used.
https://github.com/Byron/gitoxide/blob/main/gix/src/config/cache/access.rs

File: lib/src/git_backend.rs (hunk 1 above)
Comment (yuja) [reply lib/src/git_backend.rs:97]: I think both `ReadonlyGitBackend` and `MutableGitBackend` will use `gix::Repository`. Perhaps, `MutableGitBackend` wouldn't need `Mutex` as it wouldn't have to be `Sync`.

My understanding is that `gix::ThreadSafeRepository` is a pre-configured environment to instantiate a `gix::Repository` object (with some global cache?). It's something that a server preloads globally, and instantiate a repository object per user session for example.

Comment (Byron) [reply lib/src/git_backend.rs:97]: That's exactly the case, and the usage example is perfect for what `ThreadSafeRepository` and `Repository` are designed to do. This is also why `ThreadSafeRepository` can't do much more than spawn new `Repository` instances.

File: lib/testutils/src/lib.rs
Comment (Byron) [reply lib/testutils/src/lib.rs:1]: You can also run `gix config-tree` to get a list of all available (and anticipated) `git` configuration variables to learn about their usage and sometimes details or deviations.

It's true that one would configure each (`ThreadLocal`)`Repository` when opening it, and each clone will just retain that configuration. You probably want to start out with `gix::open_opts(path, gix::open::Options::isolated())`.

Comment (yuja) [reply lib/testutils/src/lib.rs:1]: > You can also run `gix config-tree` to get a list of all available (and anticipated) `git` configuration variables

Nice. Thanks for the tip.

> You probably want to start out with `gix::open_opts(path, gix::open::Options::isolated())`.

Yes, we'll probably use `Options::isolated()` to set up test fixtures. For e2e testing, I think overriding with environment variables is the way to go.

Comment (Byron) [reply lib/testutils/src/lib.rs:1]: >  I think overriding with environment variables is the way to go.

This will work under the assumption that you will use `gix::open()|gix::discover()` for that. Please also note that if you want to fetch from remotes, these options currently need to have the [`git_binary`](https://github.com/Byron/gitoxide/blob/f9ae1bc6b514b6bbcda2c672f7b7ccbd999356a0/src/plumbing/main.rs#L103) activated explicitly to pick up all the options of the git installation. Without it, `gix` won't know all credential helpers that `git` known, for example.

Comment (yuja) [reply lib/testutils/src/lib.rs:1]: > if you want to fetch from remotes, these options currently need to have the [`git_binary`](https://github.com/Byron/gitoxide/blob/f9ae1bc6b514b6bbcda2c672f7b7ccbd999356a0/src/plumbing/main.rs#L103) activated explicitly to pick up all the options of the git installation. Without it, `gix` won't know all credential helpers that `git` known, for example.

That's good to know, thanks. I'm not planned to migrate fetch/push functions anytime soon, and these functions will use git2. I don't know if libgit2 picks up the `git` installation configs, and the current situation might be the same. :)

Comment (Byron) [reply lib/testutils/src/lib.rs:1]: Oh, right, we already discussed this here as well 😅. `git2` requires extensive custom code to try to emulate credential helpers (and that's actually implemented in Rust as it's added by `git2`), and I doubt they manage to get the credential configuration of any `git` installation as that wouldn't match their policy at all (`libgit2` is standalone).
